PYME/Analysis/points/ripleys.py

```diff
@@ -106,14 +121,72 @@ def points_from_mask(mask, sampling, three_d = True, coord_origin=(0,0,0)):
         mask_area = bool_mask.sum() * vx * vy
         
     return xu, yu, zu, mask_area
-    
 
-def ripleys_k(x, y, n_bins, bin_size, mask=None, bbox=None, z=None, threaded=False, sampling=5.0, coord_origin=(0,0,0)):
+def mc_points_from_mask(mask, n_points, three_d=True, coord_origin=(0,0,0)):
+    """
+    Calculate coordinate positions in nm from a Monte-Carlo-sampled mask.
+
+    Parameters
+    ----------
+        mask : np.array
+            Mask image (array of 0s and 1s)
+        n_points : int
+            Number of points to return.
+        three_d : bool
+            Return 2D or 3D coordinates
+        coord_origin : 3-tuple
+            Offset in nm of the x, y, and z coordinates w.r.t. the camera 
+            origin (used to make sure mask aligns)
+    """
+    vx, vy, vz = mask.voxelsize
+    x0_m, y0_m, z0_m = mask.origin
+    x0_p, y0_p, z0_p = coord_origin
+
+    eps = 0.1  # scaling fudge factor
+
+    if three_d:
+        #convert mask to boolean image
+        bool_mask = np.atleast_3d(mask.data[:, :, :, 0].squeeze()) > 0.5
+        mask_area = bool_mask.sum()
+
+        # Scale up number of points to simulate so we still have n_points left 
+        # after the Monte-Carlo rejection step
+        n_sim = int((np.prod(bool_mask.shape)/mask_area + eps)*n_points)
+        
+        # generate randomly sampled coordinates on mask
+        xu, yu, zu = (np.random.rand(n_sim,3)*[bool_mask.shape[0],bool_mask.shape[1],bool_mask.shape[2]]).T
+        # Find corresponding (xu, yu, zu) in the mesj
+        point_mask = bool_mask[xu.astype(int), yu.astype(int), zu.astype(int)]
+        # Monte-Carlo reject points outside of the mask and shift points inside to coordinate position
+        xu, yu, zu = vx * xu[point_mask] + x0_m - x0_p, vy * yu[
+            point_mask] + y0_m - y0_p, vz * zu[point_mask] + z0_m - z0_p
+        
+        mask_area *= vx * vy * vz
+    else:
+        bool_mask = mask.data[:, :, :, 0].squeeze() > 0.5
+        if bool_mask.ndim > 2:
+            raise RuntimeError('Trying to calculate 2D Ripleys with 3D mask')
+        mask_area = bool_mask.sum()
+
+        n_sim = int((np.prod(bool_mask.shape)/mask_area + eps)*n_points)
+        
+        zu = None
+        xu, yu = (np.random.rand(n_sim,2)*[bool_mask.shape[0],bool_mask.shape[1]]).T
+        point_mask = bool_mask[xu.astype(int), yu.astype(int)]
+        xu, yu = vx * xu[point_mask] + x0_m - x0_p, vy * yu[point_mask] + y0_m - y0_p
+        mask_area *= vx * vy
+
+    # Truncate
+    xu, yu, zu = xu[:n_points], yu[:n_points], zu[:n_points] if zu is not None else None
+
+    return xu, yu, zu, mask_area
+
+def ripleys_k(x, y, n_bins, bin_size, mask=None, bbox=None, z=None, threaded=False, mc=False, n_sim=100, sampling=5.0, coord_origin=(0,0,0)):
```

Comment:
I think it would be a lot clearer if the MC stuff was not part of the `ripleys_k` function (it's fundamentally not part of calculating Ripleys function), but rather a, e.g., `uniform_sampling_confidence_intervals()` or similar function - I guess the advantage of putting it here as an option is that you avoid an extra `points_from_mask()` call, but it's hard to know if the diminished clarity is worth that.

Reply:
You are right. I will make a `uniform_sampling_confidence_intervals()` function. That will clean up a lot of `PYME.recipes.pointcloud.Ripleys`.